### ISSUE TITLE
Crash when JLS record fills drive

### ISSUE THREAD

**Issue by mliberty1** [jetperch/pyjoulescope_ui#185]
### Joulescope model

JS220

### UI version

1.0.15

### What OS are you seeing the problem on?

Windows 11, Windows 10

### What happened?

When I start a stream recording to a JLS file, it starts correctly.  However, if the recording ends up taking up all the remaining space on the drive, the UI crashes on closing the recording.

### What was expected?

The UI should be friendly.  It should give a warning that the drive is almost full.  It should then stop recording before the drive becomes full.  Finally, if the drive actually does become full, it should handle the condition rather than crashing the UI.

### How to reproduce?

1. Start a JLS stream recording by pressing the record button.  Record to a drive with limited remaining space.
2. Wait and monitor the drive free space.
3. When full, stop JLS streaming.  UI crashes.

**Comment by mliberty1** [comment]
The log contains lots of entries like:

```
ERROR:2023-05-17 15:35:27,687:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 15:35:27,688:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 15:35:27,688:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
WARNING:2023-05-17 15:35:27,688:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
ERROR:2023-05-17 15:35:27,690:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 15:35:27,690:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 15:35:27,690:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
WARNING:2023-05-17 15:35:27,690:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
```

However, it keeps going.  The crash appears at the end:

```
ERROR:2023-05-17 16:01:45,084:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:01:45,084:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:01:45,085:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
WARNING:2023-05-17 16:01:45,085:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
WARNING:2023-05-17 16:01:45,085:wr_fsr.c:639:pyjls.c:fsr skip: in=5782474852 expect=4205108176, skipped=1577366676
ERROR:2023-05-17 16:01:45,085:backend_win.c:83:pyjls.c:write failed 28
INFO:2023-05-17 16:01:45,085:signal_record.py:171:joulescope_ui.widgets.signal_record.signal_record.cls:on_cls_action_stop
ERROR:2023-05-17 16:01:45,085:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
INFO:2023-05-17 16:01:45,085:signal_record.py:137:joulescope_ui.widgets.signal_record.signal_record.obj:stop
ERROR:2023-05-17 16:01:45,085:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
INFO:2023-05-17 16:01:45,086:signal_record.py:146:joulescope_ui.widgets.signal_record.signal_record.obj:utc current: [5782491106, 182094700203172595]
WARNING:2023-05-17 16:01:45,086:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
INFO:2023-05-17 16:01:45,086:signal_record.py:146:joulescope_ui.widgets.signal_record.signal_record.obj:utc power: [5782485562, 182094700197219770]
WARNING:2023-05-17 16:01:45,086:wr_fsr.c:639:pyjls.c:fsr skip: in=5782474852 expect=4205108995, skipped=1577365857
INFO:2023-05-17 16:01:45,086:signal_record.py:146:joulescope_ui.widgets.signal_record.signal_record.obj:utc voltage: [5782491106, 182094700203172595]
ERROR:2023-05-17 16:02:05,092:threaded_writer.c:244:pyjls.c:flush timed out
ERROR:2023-05-17 16:02:05,092:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,093:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,093:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
WARNING:2023-05-17 16:02:06,094:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
ERROR:2023-05-17 16:02:06,094:backend_win.c:192:pyjls.c:thread close wait failed 258
ERROR:2023-05-17 16:02:06,094:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,094:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,094:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,094:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,094:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,094:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,094:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,095:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,095:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,095:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,095:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
WARNING:2023-05-17 16:02:06,095:wr_fsr.c:639:pyjls.c:fsr skip: in=5782485562 expect=4205108176, skipped=1577377386
ERROR:2023-05-17 16:02:06,095:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,096:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,096:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,096:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,097:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,096:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,097:backend_win.c:241:pyjls.c:jls_bkt_msg_unlock failed
ERROR:2023-05-17 16:02:06,097:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
WARNING:2023-05-17 16:02:06,097:threaded_writer.c:168:pyjls.c:thread msg 3:fsr returned 4:IO
ERROR:2023-05-17 16:02:06,097:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,097:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,097:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,098:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,098:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,098:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,098:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,098:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,098:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,098:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,099:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,099:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,099:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,099:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,099:backend_win.c:83:pyjls.c:write failed 28
ERROR:2023-05-17 16:02:06,100:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,100:raw.c:263:pyjls.c:error 4: jls_bk_fwrite(&self->backend, payload, hdr->payload_length)
ERROR:2023-05-17 16:02:06,100:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
ERROR:2023-05-17 16:02:06,100:raw.c:218:pyjls.c:error 4: jls_raw_wr_payload(self, hdr->payload_length, payload)
ERROR:2023-05-17 16:02:06,100:backend_win.c:216:pyjls.c:jls_bkt_msg_lock failed -1
Windows fatal exception: access violation

Thread 0x000006b0 (most recent call first):
  <no Python frame>

Thread 0x00001ca4 (most recent call first):
  File "threading.py", line 324 in wait
  File "queue.py", line 180 in get
  File "joulescope_ui\devices\jsdrv\js220.py", line 668 in _run
  File "threading.py", line 975 in run
  File "threading.py", line 1038 in _bootstrap_inner
  File "threading.py", line 995 in _bootstrap

Thread 0x00003e88 (most recent call first):
  File "joulescope_ui\main.py", line 497 in event
  File "joulescope_ui\main.py", line 631 in run
  File "joulescope_ui\entry_points\ui.py", line 47 in local_run
  File "joulescope_ui\entry_points\ui.py", line 51 in on_cmd
  File "joulescope_ui\__main__.py", line 26 in <module>
INFO:2023-05-17 16:02:06,759:js220_usb.c:1014:jsdrv:stream_in_port 21 sample_id skip: received=11608371069 expected=11607538965 d=832104, 1890 elements sent
INFO:2023-05-17 16:02:06,760:js220_usb.c:1014:jsdrv:stream_in_port 22 sample_id skip: received=11608371069 expected=11607538713 d=832356, 1764 elements sent
INFO:2023-05-17 16:02:06,760:js220_usb.c:1014:jsdrv:stream_in_port 23 sample_id skip: received=11608369275 expected=11607538713 d=830562, 7308 elements sent
INFO:2023-05-17 16:02:06,761:js220_usb.c:1014:jsdrv:stream_in_port 21 sample_id skip: received=11608372581 expected=11608371321 d=1260, 126 elements sent
INFO:2023-05-17 16:02:06,761:js220_usb.c:1014:jsdrv:stream_in_port 22 sample_id skip: received=11608372581 expected=11608371321 d=1260, 126 elements sent
INFO:2023-05-17 16:02:12,547:js220_usb.c:1014:jsdrv:stream_in_port 21 sample_id skip: received=11619945303 expected=11608573929 d=11371374, 3150 elements sent
INFO:2023-05-17 16:02:12,547:js220_usb.c:1014:jsdrv:stream_in_port 22 sample_id skip: received=11619945303 expected=11608573677 d=11371626, 3024 elements sent
INFO:2023-05-17 16:02:12,547:js220_usb.c:1014:jsdrv:stream_in_port 23 sample_id skip: received=11619943509 expected=11608573677 d=11369832, 4662 elements sent
```

**Comment by mliberty1** [comment]
Addressed in 1.0.30.
